Dim/dui3/send caching di (DUI3-79)

didimitrie opened the pull request:
DI-ifies the send conversion cache, and removes a bunch of wiggle code that was no longer needed as a result. 

Points of focus for the review is arcgis: it builds, but i haven't tested it locally to see that it does what it's supposed to. Do note the last paragraph on the cache class: 

>  Caching is optional in the send ops; an instance of this should be injected only in applications <b>where we know we can rely on change tracking!

If we do not trust change detection in arcgis (ie, difficult re object selection, layers, etc.), it can be skipped entirely! 

https://spockle.atlassian.net/browse/DUI3-29

oguzhankoral commented on DUI3-DX/Sdk/Speckle.Connectors.Utils/Caching/SendConversionCache.cs:
```diff
@@ -0,0 +1,48 @@
+///<inheritdoc/>
+public class SendConversionCache : ISendConversionCache
+{
+  public SendConversionCache() { }
```
not needed

oguzhankoral commented on DUI3-DX/DUI3/Speckle.Connectors.DUI/Models/Card/SenderModelCard.cs:
```diff
@@ -1,12 +1,11 @@
 
-  [JsonIgnore]
-  public HashSet<string> ChangedObjectIds { get; set; } = new();
+  // [JsonIgnore]
```
happy to get rid of from it. You can remove it I believe. God bless `EvictObjects`

oguzhankoral commented on DUI3-DX/Connectors/Rhino/Speckle.Connectors.Rhino7/Operations/Send/RhinoRootObjectBuilder.cs:
```diff
@@ -17,10 +19,12 @@ namespace Speckle.Connectors.Rhino7.Operations.Send;
+  private readonly ISendConversionCache? _sendConversionCache;
 
-  public RhinoRootObjectBuilder(IUnitOfWorkFactory unitOfWorkFactory)
+  public RhinoRootObjectBuilder(IUnitOfWorkFactory unitOfWorkFactory, ISendConversionCache? sendConversionCache = null)
```
If the concrete type of this interface not registered at all in Connector, this might throw but not sure. @BovineOx can put more input on this. Do we need nullability here since this is connector specific builder?

didimitrie replied:
Good point actually. If we know Rhino's going to have it, it doesn't make sense here for it to be nullable 🤔

didimitrie replied:
done, we now have a null send conversion cache for future use, and it's not optional anywhere now

oguzhankoral commented on DUI3-DX/Connectors/Revit/Speckle.Connectors.RevitShared/Bindings/RevitSendBinding.cs:
yes, thanks for it! we do it in bridge already

oguzhankoral commented on DUI3-DX/Connectors/Revit/Speckle.Connectors.RevitShared/Bindings/BasicConnectorBindingRevit.cs:
```diff
@@ -88,7 +88,7 @@ public void HighlightModel(string modelCardId)
 
     var elementIds = model.SendFilter.NotNull().GetObjectIds().Select(ElementId.Parse).ToList();
-    if (elementIds.Count != 0)
+    if (elementIds.Count == 0)
```
back to square one, I don't even know how/why it transformed that much

oguzhankoral commented on DUI3-DX/Connectors/Autocad/Speckle.Connectors.AutocadShared/HostApp/Extensions/DocumentExtensions.cs:
```diff
@@ -17,13 +18,26 @@ public static List<AutocadRootObject> GetObjects(this Document doc, IEnumerable<
           }
+          catch (Autodesk.AutoCAD.Runtime.Exception e)
+          {
+            if (e.ErrorStatus == ErrorStatus.WasErased) // Note: TBD if we want to catch more things in here. For now maybe not, but it does seem like this function gets into "crashes the host app territory"
```
Since it is in foreach loop we can maybe aggregate them and throw as global error in UI. yes, TBD

oguzhankoral commented on DUI3-DX/Connectors/Autocad/Speckle.Connectors.AutocadShared/Bindings/AutocadSendBinding.cs:
```diff
@@ -162,14 +158,14 @@ private async Task SendInternal(string modelCardId)
-
-      // It's important to reset the model card's list of changed obj ids so as to ensure we accurately keep track of changes between send operations.
-      modelCard.ChangedObjectIds = new();
+      // // Store the converted references in memory for future send operations, overwriting the existing values for the given application id.
```
I would delete as others

oguzhankoral reviewed (changes requested):
Overall I'm happy it is sorted out nicely 🙌
My main question on having caching nullable, how it can be null?
As far as I understand caching is doable in ArcGIS and it is quite important since we might have some layouts that user don't touch often. Good to cache them on send.

didimitrie commented:
> My main question on having caching nullable, how it can be null?

Basically I am not sure if all connectors will have this cache, so hence it's nullable (optinal) everywhere. Send should work without it - for example, if i comment out in rhino this line, 

``` 
    // register send conversion cache
    builder.AddSingleton<ISendConversionCache, SendConversionCache>();
```

in the DUI3-DX/Connectors/Rhino/Speckle.Connectors.Rhino7/DependencyInjection/RhinoConnectorModule.cs, it will send as always, but just not use the cache. 

Does this make sense @oguzhankoral?

KatKatKateryna commented:
As a general comment, the app is running and changedObjectIds are detected as before. 
Re change detection: ATM, we track it on layer level, and it works well. More granularity (and more complexity) can be added for report purpose, but for Selecting/Updating selection looks fine to me

BovineOx commented:
> > My main question on having caching nullable, how it can be null?
> 
> Basically I am not sure if all connectors will have this cache, so hence it's nullable (optinal) everywhere. Send should work without it - for example, if i comment out in rhino this line,
> 
> ```
>     // register send conversion cache
>     builder.AddSingleton<ISendConversionCache, SendConversionCache>();
> ```
> 
> in the DUI3-DX/Connectors/Rhino/Speckle.Connectors.Rhino7/DependencyInjection/RhinoConnectorModule.cs, it will send as always, but just not use the cache.
> 
> Does this make sense @oguzhankoral?
> 
> I think I get it now (after i replied here [#3491 (comment)](https://github.com/specklesystems/speckle-sharp/pull/3491#discussion_r1631180249)) : pretty much the only place it actually needs to be nullable is in the shared `RootObjectSender` class, which is shared by any connector, including ones that will potentially not have this cache.
> 
> That okay ?

As I mentioned above we should probably have a NoCacheSendConversionCache() I am unsure if the Singleton() is the right scope also, maybe, probably it should be as the cache will persist across sends.

JR-Morgan commented on DUI3-DX/Connectors/Revit/Speckle.Connectors.RevitShared/Bindings/RevitSendBinding.cs:
```diff
@@ -57,25 +55,16 @@ public List<ISendFilter> GetSendFilters()
 
-  public async Task Send(string modelCardId)
-  {
-    await SpeckleTopLevelExceptionHandler
```
It looks like `SpeckleTopLevelExceptionHandler` is no longer used. Shall we delete the class also

didimitrie replied:
done!

didimitrie commented:
Oki doki, ready for review again: 
- made the cache non-optional anywhere, and added a do nothing one for future use in cnx's that can't support it (cc @BovineOx @oguzhankoral )
- removed stale class and comments (cc @JR-Morgan @oguzhankoral) 

Also thx @KatKatKateryna for testing arcgis!

oguzhankoral reviewed (approved):
LGTM 🚀
